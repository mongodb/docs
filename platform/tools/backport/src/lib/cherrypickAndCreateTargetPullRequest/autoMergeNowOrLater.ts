import { ValidConfigOptions } from '../../options/options';
import { mergePullRequest } from '../github/v3/mergePullRequest';
import { GithubV4Exception } from '../github/v4/apiRequestV4';
import {
  enablePullRequestAutoMerge,
  parseGithubError,
} from '../github/v4/enablePullRequestAutoMerge';
import { logger } from '../logger';
import { ora } from '../ora';

export async function autoMergeNowOrLater(
  options: ValidConfigOptions,
  pullNumber: number,
) {
  const text = `Auto-merge: Enabling via "${options.autoMergeMethod}"`;
  logger.info(text);

  const spinner = ora(options.interactive, text).start();

  if (options.dryRun) {
    spinner.succeed();
    return;
  }

  try {
    try {
      await enablePullRequestAutoMerge(options, pullNumber);
    } catch (e) {
      if (!(e instanceof GithubV4Exception)) {
        throw e;
      }

      logger.info(
        `Auto merge: Failed to enable auto merge for PR "#${pullNumber}" due to ${e.message}`,
      );

      const { isMissingStatusChecks } = parseGithubError(e);
      if (!isMissingStatusChecks) {
        throw e;
      }

      // if auto merge cannot be enabled due to missing status checks, the PR should be merged immediately
      logger.info('Auto merge: Attempting to merge immediately');
      await mergePullRequest(options, pullNumber);
      spinner.text = 'Auto-merge: Pull request was merged immediately';
    }

    spinner.succeed();
  } catch (e) {
    logger.warn(`Auto merge: An error occurred ${e}`);
    spinner.fail();
  }
}
