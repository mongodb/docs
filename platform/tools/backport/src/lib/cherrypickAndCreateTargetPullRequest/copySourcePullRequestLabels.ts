import { ValidConfigOptions } from '../../options/options';
import { addLabelsToPullRequest } from '../github/v3/addLabelsToPullRequest';
import { Commit } from '../sourceCommit/parseSourceCommit';

export async function copySourcePullRequestLabelsToTargetPullRequest(
  options: ValidConfigOptions,
  commits: Commit[],
  pullNumber: number,
) {
  const labels = getNonBackportLabels(commits);
  if (labels.length > 0) {
    await addLabelsToPullRequest({ ...options, pullNumber, labels });
  }
}

function getNonBackportLabels(commits: Commit[]) {
  return commits.flatMap((commit) => {
    if (!commit.sourcePullRequest) {
      return [];
    }

    const backportLabels = commit.targetPullRequestStates.map((pr) => pr.label);
    const labels = commit.sourcePullRequest.labels.filter(
      (label) => !backportLabels.includes(label),
    );

    return labels;
  });
}
