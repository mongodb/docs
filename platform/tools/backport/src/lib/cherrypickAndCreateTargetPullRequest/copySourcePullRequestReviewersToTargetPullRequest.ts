import { ValidConfigOptions } from '../../options/options';
import { filterNil } from '../../utils/filterEmpty';
import { addReviewersToPullRequest } from '../github/v3/addReviewersToPullRequest';
import { getReviewersFromPullRequests } from '../github/v3/getReviewersFromPullRequests';
import { Commit } from '../sourceCommit/parseSourceCommit';

export async function copySourcePullRequestReviewersToTargetPullRequest(
  options: ValidConfigOptions,
  commits: Commit[],
  pullNumber: number,
) {
  const pullNumbers = commits
    .map((commit) => commit.sourcePullRequest?.number)
    .filter(filterNil);

  const reviewers = await getReviewersFromPullRequests({
    options,
    pullNumbers,
  });
  if (reviewers) {
    await addReviewersToPullRequest(options, pullNumber, reviewers);
  }
}
