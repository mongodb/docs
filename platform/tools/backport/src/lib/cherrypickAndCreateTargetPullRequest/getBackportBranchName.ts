import Handlebars from 'handlebars';
import { Commit } from '../../entrypoint.api';
import { ValidConfigOptions } from '../../options/options';
import { getShortSha } from '../github/commitFormatters';

/*
 * Returns the name of the backport branch without remote name
 *
 * Examples:
 * For a single PR: `backport/7.x/pr-1234`
 * For a single commit: `backport/7.x/commit-abcdef`
 * For multiple: `backport/7.x/pr-1234_commit-abcdef`
 */
export function getBackportBranchName({
  options,
  targetBranch,
  commits,
}: {
  options: ValidConfigOptions;
  targetBranch: string;
  commits: Commit[];
}) {
  const refValues = commits
    .map((commit) =>
      commit.sourcePullRequest
        ? `pr-${commit.sourcePullRequest.number}`
        : `commit-${getShortSha(commit.sourceCommit.sha)}`,
    )
    .join('_')
    .slice(0, 200);

  const sourcePullRequest = commits[0].sourcePullRequest; // assume that all commits are from the same PR
  const defaultBackportBranchName = 'backport/{{targetBranch}}/{{refValues}}';

  const backportBranchName =
    options.backportBranchName ?? defaultBackportBranchName;

  const template = Handlebars.compile(backportBranchName);
  return template({
    sourcePullRequest,
    targetBranch,
    refValues,
  });
}
