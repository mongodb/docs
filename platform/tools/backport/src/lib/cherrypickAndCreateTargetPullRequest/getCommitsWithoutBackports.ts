import chalk from 'chalk';
import { Commit } from '../../entrypoint.api';
import { ValidConfigOptions } from '../../options/options';
import { getIsCommitInBranch } from '../git';
import { getFirstLine } from '../github/commitFormatters';
import { fetchCommitsByAuthor } from '../github/v4/fetchCommits/fetchCommitsByAuthor';

// when the user is facing a git conflict we should help them understand
// why the conflict occurs. In many cases it's because one or more commits haven't been backported yet
export async function getCommitsWithoutBackports({
  options,
  // commit that is being backported
  commit,

  // branch the commit is being backported to
  targetBranch,

  // the files with conflicts on the target branch
  conflictingFiles,
}: {
  options: ValidConfigOptions;
  commit: Commit;
  targetBranch: string;
  conflictingFiles: string[];
}) {
  // commits on the source branch that also touched the conflicting files
  const commitsInConflictingPaths = await fetchCommitsByAuthor({
    ...options,
    author: null, // retrieve commits across all authors
    maxNumber: 50,
    dateSince: null,
    dateUntil: commit.sourceCommit.committedDate,
    commitPaths: conflictingFiles,
  });

  const promises = await Promise.all(
    commitsInConflictingPaths
      .filter((c) => {
        // exclude the commit we are currently trying to backport
        if (c.sourceCommit.sha === commit.sourceCommit.sha) {
          return false;
        }

        // exclude commits that are newer than the commit we are trying to backport
        if (c.sourceCommit.committedDate > commit.sourceCommit.committedDate) {
          return false;
        }

        // only consider commits that have an associated pull request
        if (!c.sourcePullRequest?.url) {
          return false;
        }

        // only include commit if it has an unmerged PR for the given target branch
        const hasUnmergedPr = c.targetPullRequestStates.some(
          (pr) => pr.branch === targetBranch && pr.state !== 'MERGED',
        );

        return hasUnmergedPr;
      })
      .slice(0, 10) // limit to max 10 commits
      .map(async (c) => {
        const results = await Promise.all(
          c.targetPullRequestStates.map(async (targetPr) => {
            if (!targetPr.mergeCommit) {
              return false;
            }

            return getIsCommitInBranch(options, targetPr.mergeCommit.sha);
          }),
        );

        const isCommitInBranch = results.some((inBranch) => inBranch === true);
        return { c, isCommitInBranch };
      }),
  );

  return promises
    .filter(({ isCommitInBranch }) => !isCommitInBranch)
    .map(({ c }) => {
      // get pull request for the target branch (if it exists)
      const pendingBackportPr = c.targetPullRequestStates.find(
        (pr) => pr.branch === targetBranch && pr.state === 'OPEN',
      );

      const formatted = pendingBackportPr
        ? ` - ${getFirstLine(c.sourceCommit.message)} ${chalk.gray(
            '(backport pending)',
          )}\n   ${pendingBackportPr.url}`
        : ` - ${getFirstLine(c.sourceCommit.message)} ${chalk.red(
            '(backport missing)',
          )}\n   ${c.sourcePullRequest?.url}`;

      return { formatted, commit: c };
    });
}
