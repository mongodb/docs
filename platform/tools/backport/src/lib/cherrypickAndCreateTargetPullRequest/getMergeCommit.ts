import { Commit } from '../../entrypoint.api';
import { ValidConfigOptions } from '../../options/options';
import { getIsMergeCommit, getShasInMergeCommit } from '../git';
import { fetchCommitBySha } from '../github/v4/fetchCommits/fetchCommitBySha';

export async function getMergeCommits(
  options: ValidConfigOptions,
  commit: Commit,
): Promise<Commit[]> {
  const { sha } = commit.sourceCommit;
  if (!options.mainline) {
    const isMergeCommit = await getIsMergeCommit(options, sha);
    if (isMergeCommit) {
      const shas = await getShasInMergeCommit(options, sha);
      return Promise.all(
        shas.reverse().map((sha) => fetchCommitBySha({ ...options, sha })),
      );
    }
  }

  return [commit];
}
