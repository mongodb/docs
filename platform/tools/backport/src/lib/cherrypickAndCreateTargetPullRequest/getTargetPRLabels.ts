import { uniq } from 'lodash';
import { Commit } from '../../entrypoint.api';
import { filterNil } from '../../utils/filterEmpty';
import { getSourceBranchFromCommits } from '../getSourceBranchFromCommits';
import { logger } from '../logger';

export function getTargetPRLabels({
  interactive,
  targetPRLabels,
  commits,
  targetBranch,
}: {
  interactive: boolean;
  targetPRLabels: string[];
  commits: Commit[];
  targetBranch: string;
}) {
  const labels = getLabels({
    commits,
    targetBranch,
    targetPRLabels,
    interactive,
  });

  return uniq(labels);
}

function getLabels({
  commits,
  targetBranch,
  targetPRLabels,
  interactive,
}: {
  commits: Commit[];
  targetBranch: string;
  targetPRLabels: string[];
  interactive: boolean;
}) {
  const sourceBranch = getSourceBranchFromCommits(commits);
  const labels = commits
    .flatMap((c) => {
      const targetPullRequest = c.targetPullRequestStates.find(
        (pr) => pr.branch === targetBranch,
      );

      if (!targetPullRequest?.branchLabelMappingKey) {
        logger.info('Missing branchLabelMappingKey for target pull request');

        // remove dynamic labels like `$1` in interactive mode
        return targetPRLabels.filter((l) => {
          return l.match(/\$\d/) === null || !interactive;
        });
      }

      const regex = new RegExp(targetPullRequest.branchLabelMappingKey);

      return targetPRLabels.map((targetPRLabel) => {
        return targetPullRequest.label?.replace(regex, targetPRLabel);
      });
    })
    .filter(filterNil)
    .map((label) => {
      return label
        .replaceAll('{{targetBranch}}', targetBranch)
        .replaceAll('{{sourceBranch}}', sourceBranch);
    });

  return labels;
}
