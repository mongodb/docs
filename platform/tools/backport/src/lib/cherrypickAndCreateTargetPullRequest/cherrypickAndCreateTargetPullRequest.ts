import chalk from 'chalk';
import { flatten } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import { getSourceBranchFromCommits } from '../getSourceBranchFromCommits';
import {
  createBackportBranch,
  deleteBackportBranch,
  pushBackportBranch,
  getRepoForkOwner,
} from '../git';
import { addAssigneesToPullRequest } from '../github/v3/addAssigneesToPullRequest';
import { addLabelsToPullRequest } from '../github/v3/addLabelsToPullRequest';
import { addReviewersToPullRequest } from '../github/v3/addReviewersToPullRequest';
import {
  createPullRequest,
  PullRequestPayload,
} from '../github/v3/getPullRequest/createPullRequest';
import { getPullRequestBody } from '../github/v3/getPullRequest/getPullRequestBody';
import { getTitle } from '../github/v3/getPullRequest/getTitle';
import { validateTargetBranch } from '../github/v4/validateTargetBranch';
import { consoleLog } from '../logger';
import { Target } from '../runSequentially';
import { sequentially } from '../sequentially';
import { Commit } from '../sourceCommit/parseSourceCommit';
import { autoMergeNowOrLater } from './autoMergeNowOrLater';
import { copySourcePullRequestLabelsToTargetPullRequest } from './copySourcePullRequestLabels';
import { copySourcePullRequestReviewersToTargetPullRequest } from './copySourcePullRequestReviewersToTargetPullRequest';
import { getBackportBranchName } from './getBackportBranchName';
import { getMergeCommits } from './getMergeCommit';
import { getTargetPRLabels } from './getTargetPRLabels';
import { waitForCherrypick } from './waitForCherrypick';

export async function cherrypickAndCreateTargetPullRequest({
  options,
  commits,
  target,
}: {
  options: ValidConfigOptions;
  commits: Commit[];
  target: Target;
}): Promise<{ url: string; number: number; didUpdate: boolean }> {
  const { branch: targetBranch } = target;

  const backportBranch = getBackportBranchName({
    options,
    targetBranch,
    commits,
  });

  // TODO: backportBranch must start with a valid character

  const repoForkOwner = getRepoForkOwner(options);
  consoleLog(`\n${chalk.bold(`Backporting to ${JSON.stringify(target)}:`)}`);

  await validateTargetBranch({ ...options, branchName: targetBranch });
  await createBackportBranch({
    options,
    sourceBranch: getSourceBranchFromCommits(commits),
    targetBranch,
    backportBranch,
  });

  const commitsFlattened = flatten(
    await Promise.all(commits.map((c) => getMergeCommits(options, c))),
  );

  const cherrypickResults = await sequentially(commitsFlattened, (commit) =>
    waitForCherrypick(options, commit, target),
  );
  const hasAnyCommitWithConflicts = cherrypickResults.some(
    (r) => r.hasCommitsWithConflicts,
  );

  if (!options.dryRun) {
    await pushBackportBranch({ options, backportBranch });
    await deleteBackportBranch({ options, backportBranch });
  }

  const prPayload: PullRequestPayload = {
    owner: options.repoOwner,
    repo: options.repoName,
    title: getTitle({ options, commits, targetBranch }),
    body: getPullRequestBody({ options, commits, targetBranch }),
    head: `${repoForkOwner}:${backportBranch}`, // eg. sorenlouv:backport/7.x/pr-75007
    base: targetBranch, // eg. 7.x
    draft: options.draft,
  };

  const targetPullRequest = await createPullRequest({ options, prPayload });

  // add assignees to target pull request
  const assignees = options.autoAssign
    ? [options.authenticatedUsername]
    : options.assignees;

  if (options.assignees.length > 0) {
    await addAssigneesToPullRequest({
      ...options,
      pullNumber: targetPullRequest.number,
      assignees,
    });
  }

  // add reviewers to target pull request
  if (options.reviewers.length > 0) {
    await addReviewersToPullRequest(
      options,
      targetPullRequest.number,
      options.reviewers,
    );
  }

  // add reviewers of the original PRs to the target pull request
  if (options.copySourcePRReviewers) {
    await copySourcePullRequestReviewersToTargetPullRequest(
      options,
      commits,
      targetPullRequest.number,
    );
  }

  // add labels to target pull request
  if (options.targetPRLabels.length > 0) {
    const labels = getTargetPRLabels({
      interactive: options.interactive,
      targetPRLabels: options.targetPRLabels,
      commits,
      targetBranch,
    });

    await addLabelsToPullRequest({
      ...options,
      pullNumber: targetPullRequest.number,
      labels,
    });
  }

  if (options.copySourcePRLabels) {
    await copySourcePullRequestLabelsToTargetPullRequest(
      options,
      commits,
      targetPullRequest.number,
    );
  }

  // make PR auto mergable
  if (options.autoMerge && !hasAnyCommitWithConflicts) {
    await autoMergeNowOrLater(options, targetPullRequest.number);
  }

  // add labels to source pull requests
  if (options.sourcePRLabels.length > 0) {
    const promises = commits.map((commit) => {
      if (commit.sourcePullRequest) {
        return addLabelsToPullRequest({
          ...options,
          pullNumber: commit.sourcePullRequest.number,
          labels: options.sourcePRLabels,
        });
      }
    });

    await Promise.all(promises);
  }

  consoleLog(`View pull request: ${targetPullRequest.url}`);

  return targetPullRequest;
}
