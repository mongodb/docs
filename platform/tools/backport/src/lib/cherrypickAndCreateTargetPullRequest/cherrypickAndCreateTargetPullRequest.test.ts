import os from 'os';
import nock from 'nock';
import { ValidConfigOptions } from '../../options/options';
import {
  listenForCallsToNockScope,
  mockGqlRequest,
} from '../../test/nockHelpers';
import { SpyHelper } from '../../types/SpyHelper';
import * as childProcess from '../child-process-promisified';
import { TargetBranchResponse } from '../github/v4/validateTargetBranch';
import * as logger from '../logger';
import * as oraModule from '../ora';
import { Commit } from '../sourceCommit/parseSourceCommit';
import * as autoMergeNowOrLater from './autoMergeNowOrLater';
import { cherrypickAndCreateTargetPullRequest } from './cherrypickAndCreateTargetPullRequest';

describe('cherrypickAndCreateTargetPullRequest', () => {
  let execSpy: SpyHelper<typeof childProcess.spawnPromise>;
  let addLabelsScope: ReturnType<typeof nock>;
  let consoleLogSpy: SpyHelper<(typeof logger)['consoleLog']>;
  let autoMergeSpy: SpyHelper<typeof autoMergeNowOrLater.autoMergeNowOrLater>;

  beforeEach(() => {
    jest.spyOn(os, 'homedir').mockReturnValue('/myHomeDir');

    autoMergeSpy = jest.spyOn(autoMergeNowOrLater, 'autoMergeNowOrLater');

    execSpy = jest
      .spyOn(childProcess, 'spawnPromise')

      // mock all spawn commands to respond without errors
      .mockResolvedValue({ stdout: '', stderr: '', code: 0, cmdArgs: [] });

    consoleLogSpy = jest.spyOn(logger, 'consoleLog');

    // ensure labels are added
    addLabelsScope = nock('https://api.github.com')
      .post('/repos/elastic/kibana/issues/1337/labels', {
        labels: ['backport'],
      })
      .reply(200);
  });

  afterEach(() => {
    jest.clearAllMocks();
    addLabelsScope.done();
    nock.cleanAll();
  });

  describe('when commit has a pull request reference', () => {
    let res: Awaited<ReturnType<typeof cherrypickAndCreateTargetPullRequest>>;
    let createPullRequestCalls: unknown[];
    let oraSpy: jest.SpyInstance;

    beforeEach(async () => {
      const options = {
        assignees: [] as string[],
        authenticatedUsername: 'sqren_authenticated',
        author: 'sorenlouv',
        autoMerge: true,
        autoMergeMethod: 'squash',
        fork: true,
        gitAuthorEmail: 'soren@louv.dk',
        gitAuthorName: 'Soren L',
        githubApiBaseUrlV4: 'http://localhost/graphql',
        interactive: false,
        prTitle: '[{{targetBranch}}] {{commitMessages}}',
        repoForkOwner: 'sorenlouv',
        repoName: 'kibana',
        repoOwner: 'elastic',
        reviewers: [] as string[],
        sourceBranch: 'myDefaultSourceBranch',
        sourcePRLabels: [] as string[],
        targetPRLabels: ['backport'],
      } as ValidConfigOptions;

      const commits: Commit[] = [
        {
          author: {
            email: 'soren.louv@elastic.co',
            name: 'Søren Louv-Jansen',
          },
          sourceBranch: '7.x',
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: {},
            committedDate: 'fff',
            sha: 'mySha',
            message: 'My original commit message (#1000)',
          },
          sourcePullRequest: {
            labels: [],
            url: 'foo',
            number: 1000,
            title: 'My original commit message',
            mergeCommit: {
              sha: 'mySha',
              message: 'My original commit message (#1000)',
            },
          },
          targetPullRequestStates: [],
        },
        {
          author: {
            email: 'soren.louv@elastic.co',
            name: 'Søren Louv-Jansen',
          },
          sourceBranch: '7.x',
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: {},
            committedDate: 'ggg',
            sha: 'mySha2',
            message: 'My other commit message (#2000)',
          },
          sourcePullRequest: {
            labels: [],
            url: 'foo',
            number: 2000,
            title: 'My other commit message',
            mergeCommit: {
              sha: 'mySha2',
              message: 'My other commit message (#2000)',
            },
          },
          targetPullRequestStates: [],
        },
      ];

      mockGqlRequest<TargetBranchResponse>({
        name: 'GetBranchId',
        statusCode: 200,
        body: { data: { repository: { ref: { id: 'foo' } } } },
      });

      const scope = nock('https://api.github.com')
        .post('/repos/elastic/kibana/pulls')
        .reply(200, { number: 1337, html_url: 'myHtmlUrl' });
      createPullRequestCalls = listenForCallsToNockScope(scope);

      oraSpy = jest.spyOn(oraModule, 'ora');

      res = await cherrypickAndCreateTargetPullRequest({
        options,
        commits,
        targetBranch: '6.x',
      });

      scope.done();
      nock.cleanAll();
    });

    it('creates the pull request with multiple PR references', () => {
      expect(createPullRequestCalls).toMatchInlineSnapshot(`
        [
          {
            "base": "6.x",
            "body": "# Backport

        This will backport the following commits from \`7.x\` to \`6.x\`:
         - [My original commit message (#1000)](foo)
         - [My other commit message (#2000)](foo)

        <!--- Backport version: 1.2.3-mocked -->

        ### Questions ?
        Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)",
            "head": "sorenlouv:backport/6.x/pr-1000_pr-2000",
            "title": "[6.x] My original commit message (#1000) | My other commit message (#2000)",
          },
        ]
      `);
    });

    it('calls autoMergeNowOrLater', () => {
      expect(autoMergeSpy).toHaveBeenCalledWith(expect.any(Object), 1337);
    });

    it('returns the expected response', () => {
      expect(res).toEqual({ didUpdate: false, url: 'myHtmlUrl', number: 1337 });
    });

    it('should make correct git commands', () => {
      expect(execSpy.mock.calls).toMatchSnapshot();
    });

    it('logs correctly', () => {
      expect(consoleLogSpy.mock.calls.length).toBe(2);
      expect(consoleLogSpy.mock.calls[0][0]).toMatchInlineSnapshot(`
        "
        Backporting to 6.x:"
      `);
      expect(consoleLogSpy.mock.calls[1][0]).toMatchInlineSnapshot(
        `"View pull request: myHtmlUrl"`,
      );
    });

    it('should start the spinner with the correct text', () => {
      expect(oraSpy.mock.calls.map(([, text]) => text)).toMatchInlineSnapshot(`
[
  "",
  "Pulling latest changes",
  "Cherry-picking: My original commit message (#1000)",
  "Cherry-picking: My other commit message (#2000)",
  "Pushing branch "sorenlouv:backport/6.x/pr-1000_pr-2000"",
  undefined,
  "Creating pull request",
  "Adding labels: backport",
  "Auto-merge: Enabling via "squash"",
]
`);
    });
  });

  describe('when commit does not have a pull request reference', () => {
    let res: Awaited<ReturnType<typeof cherrypickAndCreateTargetPullRequest>>;
    let createPullRequestCalls: unknown[];

    beforeEach(async () => {
      const options = {
        assignees: [] as string[],
        authenticatedUsername: 'sqren_authenticated',
        author: 'sorenlouv',
        fork: true,
        prTitle: '[{{targetBranch}}] {{commitMessages}}',
        repoForkOwner: 'the_fork_owner',
        repoName: 'kibana',
        repoOwner: 'elastic',
        reviewers: [] as string[],
        sourcePRLabels: [] as string[],
        targetPRLabels: ['backport'],
        githubApiBaseUrlV4: 'http://localhost/graphql',
      } as ValidConfigOptions;

      const commits: Commit[] = [
        {
          author: {
            email: 'soren.louv@elastic.co',
            name: 'Søren Louv-Jansen',
          },
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: {},
            committedDate: 'hhh',
            sha: 'mySha',
            message: 'My original commit message',
          },
          sourceBranch: '7.x',
          targetPullRequestStates: [],
        },
      ];

      mockGqlRequest<TargetBranchResponse>({
        name: 'GetBranchId',
        statusCode: 200,
        body: { data: { repository: { ref: { id: 'foo' } } } },
      });

      const scope = nock('https://api.github.com')
        .post('/repos/elastic/kibana/pulls')
        .reply(200, { number: 1337, html_url: 'myHtmlUrl' });

      createPullRequestCalls = listenForCallsToNockScope(scope);

      res = await cherrypickAndCreateTargetPullRequest({
        options,
        commits,
        targetBranch: '6.x',
      });
      scope.done();
      nock.cleanAll();
    });

    it('creates the pull request with commit reference', () => {
      expect(createPullRequestCalls).toMatchInlineSnapshot(`
        [
          {
            "base": "6.x",
            "body": "# Backport

        This will backport the following commits from \`7.x\` to \`6.x\`:
         - My original commit message (mySha)

        <!--- Backport version: 1.2.3-mocked -->

        ### Questions ?
        Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)",
            "head": "the_fork_owner:backport/6.x/commit-mySha",
            "title": "[6.x] My original commit message",
          },
        ]
      `);
    });

    it('returns the expected response', () => {
      expect(res).toEqual({ didUpdate: false, url: 'myHtmlUrl', number: 1337 });
    });
  });

  describe('when cherry-picking fails', () => {
    let res: Awaited<ReturnType<typeof cherrypickAndCreateTargetPullRequest>>;
    let createPullRequestCalls: unknown[];

    beforeEach(async () => {
      const options = {
        assignees: [] as string[],
        authenticatedUsername: 'sqren_authenticated',
        author: 'sorenlouv',
        fork: true,
        githubApiBaseUrlV4: 'http://localhost/graphql',
        prTitle: '[{{targetBranch}}] {{commitMessages}}',
        repoForkOwner: 'sorenlouv',
        repoName: 'kibana',
        repoOwner: 'elastic',
        reviewers: [] as string[],
        sourceBranch: 'myDefaultSourceBranch',
        sourcePRLabels: [] as string[],
        targetPRLabels: ['backport'],
      } as ValidConfigOptions;

      mockGqlRequest<TargetBranchResponse>({
        name: 'GetBranchId',
        statusCode: 200,
        body: { data: { repository: { ref: { id: 'foo' } } } },
      });

      const scope = nock('https://api.github.com')
        .post('/repos/elastic/kibana/pulls')
        .reply(200, { number: 1337, html_url: 'myHtmlUrl' });

      createPullRequestCalls = listenForCallsToNockScope(scope);

      res = await cherrypickAndCreateTargetPullRequest({
        options,
        commits: [
          {
            author: {
              email: 'soren.louv@elastic.co',
              name: 'Søren Louv-Jansen',
            },
            suggestedTargetBranches: [],
            sourceCommit: {
              branchLabelMapping: {},
              committedDate: '2021-08-18T16:11:38Z',
              sha: 'mySha',
              message: 'My original commit message',
            },
            sourceBranch: '7.x',
            targetPullRequestStates: [],
          },
        ],
        targetBranch: '6.x',
      });

      scope.done();
      nock.cleanAll();
    });

    it('creates the pull request with commit reference', () => {
      expect(createPullRequestCalls).toMatchInlineSnapshot(`
        [
          {
            "base": "6.x",
            "body": "# Backport

        This will backport the following commits from \`7.x\` to \`6.x\`:
         - My original commit message (mySha)

        <!--- Backport version: 1.2.3-mocked -->

        ### Questions ?
        Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)",
            "head": "sorenlouv:backport/6.x/commit-mySha",
            "title": "[6.x] My original commit message",
          },
        ]
      `);
    });

    it('returns the expected response', () => {
      expect(res).toEqual({ didUpdate: false, url: 'myHtmlUrl', number: 1337 });
    });
  });
});
