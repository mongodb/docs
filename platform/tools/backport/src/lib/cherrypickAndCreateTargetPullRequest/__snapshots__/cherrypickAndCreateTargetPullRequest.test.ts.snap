// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cherrypickAndCreateTargetPullRequest when commit has a pull request reference should make correct git commands 1`] = `
[
  [
    "git",
    [
      "reset",
      "--hard",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "clean",
      "-d",
      "--force",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "checkout",
      "-B",
      "__backport_tool_tmp",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "fetch",
      "elastic",
      "6.x:6.x",
      "--force",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "checkout",
      "-B",
      "backport/6.x/pr-1000_pr-2000",
      "elastic/6.x",
      "--no-track",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "branch",
      "-D",
      "__backport_tool_tmp",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "fetch",
      "elastic",
      "7.x:7.x",
      "--force",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "rev-list",
      "-1",
      "--merges",
      "mySha~1..mySha",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "rev-list",
      "-1",
      "--merges",
      "mySha2~1..mySha2",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "-c",
      "user.name="Soren L"",
      "-c",
      "user.email="soren@louv.dk"",
      "cherry-pick",
      "-x",
      "mySha",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "-c",
      "user.name="Soren L"",
      "-c",
      "user.email="soren@louv.dk"",
      "commit",
      "--no-edit",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "-c",
      "user.name="Soren L"",
      "-c",
      "user.email="soren@louv.dk"",
      "cherry-pick",
      "-x",
      "mySha2",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "-c",
      "user.name="Soren L"",
      "-c",
      "user.email="soren@louv.dk"",
      "commit",
      "--no-edit",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "push",
      "sorenlouv",
      "backport/6.x/pr-1000_pr-2000:backport/6.x/pr-1000_pr-2000",
      "--force",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "reset",
      "--hard",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "checkout",
      "myDefaultSourceBranch",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
  [
    "git",
    [
      "branch",
      "--delete",
      "--force",
      "backport/6.x/pr-1000_pr-2000",
    ],
    "/myHomeDir/.backport/repositories/elastic/kibana",
  ],
]
`;
